adapters/silverpush/silverpush.go

```diff
@@ -0,0 +1,344 @@
+package silverpush
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	bidderConfig  = "sp_pb_ortb"
+	bidderVersion = "1.0.0"
+)
+
+type adapter struct {
+	bidderName string
+	endpoint   string
+}
+
+type SilverPushImpExt map[string]json.RawMessage
+
+type SilverPushReqExt struct {
+	PublisherId string  `json:"publisherId"`
+	BidFloor    float64 `json:"bidfloor"`
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	imps := request.Imp
+	var errors []error
+	requests := make([]*adapters.RequestData, 0, len(imps))
+
+	for _, imp := range imps {
+		impsByMediaType, err := impressionByMediaType(&imp)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		for _, impByMediaType := range impsByMediaType {
+			request.Imp = []openrtb2.Imp{impByMediaType}
+
+			if err := validateRequest(request); err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			requestData, err := a.makeRequest(request)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+
+			requests = append(requests, requestData)
+		}
+	}
+	return requests, errors
+}
+
+func (a *adapter) makeRequest(req *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	reqJSON, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}, nil
+}
+
+func validateRequest(req *openrtb2.BidRequest) error {
+	imp := &req.Imp[0]
+	var silverPushExt openrtb_ext.ImpExtSilverpush
+	err := setPublisherId(req, imp, &silverPushExt)
+	if err != nil {
+		return err
+	}
+	if err := setUser(req); err != nil {
+		return err
+	}
+	setDevice(req)
+
+	err = setExtToRequest(req, silverPushExt.PublisherId)
+	if err != nil {
+		return err
+	}
+	return setImpForAdExchange(imp, &silverPushExt)
+}
+
+func setDevice(req *openrtb2.BidRequest) {
+	if req.Device != nil {
+		deviceCopy := *req.Device
+		if len(deviceCopy.UA) > 0 {
+			deviceCopy.OS = getOS(deviceCopy.UA)
+			if isMobile(deviceCopy.UA) {
+				deviceCopy.DeviceType = 1
+			} else if isCTV(deviceCopy.UA) {
+				deviceCopy.DeviceType = 3
+			} else {
+				deviceCopy.DeviceType = 2
+			}
+		}
+		req.Device = &deviceCopy
+	}
+}
```

Comment:
Deeply nested `if` statements are difficult to follow. Please consider inverting the checks in this method and `setUser`. Example:

```
func setDevice(req *openrtb2.BidRequest) {
    if req.Device == nil {
        return
    }

    if len(deviceCopy.UA) == 0 {
        return
    }

    deviceCopy := *req.Device
    deviceCopy.OS = getOS(deviceCopy.UA)

    if isMobile(deviceCopy.UA) {
        deviceCopy.DeviceType = 1
    } else if isCTV(deviceCopy.UA) {
        deviceCopy.DeviceType = 3
    } else {
        deviceCopy.DeviceType = 2
    }

    req.Device = &deviceCopy
}
```

Reply:
Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)